### PR TITLE
Refactor rpc

### PR DESCRIPTION
**Pull request by zhongqiuwood**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Review by AaronZgl — DISMISSED** [review]

File: x/evm/watcher/types.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	return m.JsonEthTx
+}
+
+type MsgCode struct {
+	Key  string
+	Code string
+}
+
+type CodeInfo struct {
+	Height uint64 `height`
```
Comment (xiangjianmeng) [inline x/evm/watcher/types.go]: should be `json:"height"`

File: x/evm/handler.go [hunk 2]
```diff
@@ -104,15 +105,22 @@ func handleMsgEthereumTx(ctx sdk.Context, k *Keeper, msg types.MsgEthereumTx) (*
...
+			k.Watcher.SaveTransactionReceipt(0, msg, common.BytesToHash(txHash), uint64(k.TxCount-1), &types.ResultData{}, ctx.GasMeter().GasConsumed())
+		}
 		return nil, err
 	}
 
 	if !st.Simulate {
 		// update block bloom filter
 		k.Bloom.Or(k.Bloom, executionResult.Bloom)
 		k.LogSize = st.Csdb.GetLogSize()
+		k.Watcher.SaveTransactionReceipt(1, msg, common.BytesToHash(txHash), uint64(k.TxCount-1), resultData, ctx.GasMeter().GasConsumed())
```
Comment (xiangjianmeng) [inline x/evm/handler.go]: not  use magic number

File: x/evm/watcher/db.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"path/filepath"
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/spf13/viper"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+const ViperFlagFastQuery = "fast-query"
```
Comment (xiangjianmeng) [inline x/evm/watcher/db.go]: repeat with FlagFastQuery in cmd.client

**Review by xiangjianmeng — DISMISSED** [review]

**Comment by KamiD** [comment]
Code changes:
* Disable the mutex in function of `abci.QuerySync` by flag `close-mutex`, default value is true(means mutex used)
* Add callback in x/evm to watch the execution of the evm transactions and save result to the level-db which was named `watcher.db`
* Add `watcher.db` to refactor query data source of the following RPC interfaces(don't need to query by function `QuerySync` anymore):
   - eth_blockNumber
   - eth_getCode
   - eth_getBlockByHash
   - eth_getBlockByNumber
   - eth_getTransactionByHash
   - eth_getTransactionByBlockHashAndIndex
   - eth_getTransactionReceipt
* Add flag `fast-query` to control the switch of `watcher.db`, default value is false(means fast-query closed)

**Review by xiangjianmeng — APPROVED** [review]

**Review by summerpro — APPROVED** [review]

**Review by AaronZgl — APPROVED** [review]

**Comment by summerpro** [comment]
It is recommended to use rlp or protobuf3 encoding in the future, which should be smaller than the stored data of json.Marshal

**Pull request merged**